pysph/sph/rigid_body.py

```diff
@@ -225,13 +225,41 @@ def initialize(self, d_idx, d_m, d_fx, d_fy, d_fz):
         d_fz[d_idx] = d_m[d_idx]*self.gz
 
 
+class SummationDensityBoundary(Equation):
+    r"""Good old Summation density:
+
+    :math:`\rho_a = \sum_b m_b W_{ab}`
+
+    """
+    def __init__(self, dest, sources, fluid_rho=1000.0):
+        self.fluid_rho = fluid_rho
+        super(SummationDensityBoundary, self).__init__(dest, sources)
+
+    def loop(self, d_idx, d_rho, s_idx, s_m, s_V, WIJ):
+        d_rho[d_idx] += self.fluid_rho * s_V[s_idx] * WIJ
+
+
+class CutoffDensity(Equation):
+    """Limit density to real fluid density
+
+
+    """
+    def __init__(self, dest, sources, fluid_rho=1000.0):
+        self.fluid_rho = fluid_rho
+        super(CutoffDensity, self).__init__(dest, sources)
+
+    def initialize(self, d_idx, d_rho):
```

Comment:
Is this really needed?  Instead of changing the density directly, you should probably have a different equation of state that does not generate negative pressures by finding `max(rho0, d_rho[d_idx))` when computing the pressure.  Otherwise your `m_j/rho_j` terms will all be wrong.  So in short, don't change the rho but use a different EOS.